src/server/pps/server/pipeline_controller.go

```diff
@@ -145,19 +148,35 @@ func (a *apiServer) step(pachClient *client.APIClient, pipeline string, keyVer,
 			if err := op.scaleUpPipeline(); err != nil {
 				return err
 			}
-			return op.setPipelineState(pps.PipelineState_PIPELINE_RUNNING)
+			return op.setPipelineState(pps.PipelineState_PIPELINE_RUNNING, "")
 		}
 		// don't want cron commits or STANDBY state changes while pipeline is
 		// stopped
 		op.stopPipelineMonitor()
 		// default: scale down if pause/standby hasn't propagated to etcd yet
 		return op.scaleDownPipeline()
 	case pps.PipelineState_PIPELINE_FAILURE:
-		// pipeline fails if docker image isn't found
+		// pipeline fails if it encounters an unrecoverable error
 		if err := op.finishPipelineOutputCommits(); err != nil {
 			return err
 		}
 		return op.deletePipelineResources()
+	case pps.PipelineState_PIPELINE_CRASHING:
+		if !op.rcIsFresh() {
+			return op.restartPipeline("stale RC") // step() will be called again after etcd write
+		}
+		if op.pipelineInfo.Stopped {
+			return op.setPipelineState(pps.PipelineState_PIPELINE_PAUSED, "")
+		}
+		workersUp, err := a.allWorkersUp(pachClient, op.pipelineInfo)
+		if err != nil {
+			return err
+		}
+		if workersUp {
+			return op.setPipelineState(pps.PipelineState_PIPELINE_RUNNING, "")
+		}
+		time.Sleep(crashingBackoff)
+		return op.setPipelineState(pps.PipelineState_PIPELINE_CRASHING, op.pipelineInfo.Reason)
```

Comment:
There's a prior response with more detail (which I can move here if it would be helpful) but I think this could be moved into a separate goro, which would shorten the amount of work done in the critical path of the controller and remove the feedback loop created by overwriting the pipeline state to `CRASHING`.

Also, it occurs to me that setting the pipeline state after a long sleep could create a race where a user updates a pipeline (putting it into `STARTING`) during the 10-second sleep, and then the new pipeline is transistioned straight into `CRASHING` when the sleep terminates, even if the new pipeline may be healthy.

Reply:
That makes a lot of sense, to make sure I understand what you're saying here: This would spin off a goro that polls k8s at some interval to see if the pods are up, and if they are it would transition the state into running and exit? That sounds a lot better to me.

Reply:
Yup, exactly (and you might be alluding to this already, but there's even a function, transitionPipelineState, that the goro can call to make sure there wasn't a `stopPipeline` call made while the pipeline was in `CRASHING` that would be undone by the transition to `RUNNING`)